[AOTI] Emit a CMakeLists.txt when package_cpp_only

desertfire opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #143352

Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @chauhang @aakhundov

Differential Revision: [D67458526](https://our.internmc.facebook.com/intern/diff/D67458526)

desertfire committed:
[AOTI] Emit a CMakeLists.txt when package_cpp_only

Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Emit a CMakeLists.txt when package_cpp_only"


Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Emit a CMakeLists.txt when package_cpp_only"


Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire committed:
Update on "[AOTI] Emit a CMakeLists.txt when package_cpp_only"


Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

malfet commented on test/inductor/test_aot_inductor_package.py:
```diff
@@ -176,6 +180,60 @@ def forward(self, x, y):
         )
         self.check_model(Model(), example_inputs)
 
+    def test_compile_after_package(self):
```
It would be nice to skip skip tests if `cmake` or `make` could not be found on the tester

malfet commented on torch/_inductor/cpp_builder.py:
```diff
@@ -1545,3 +1547,65 @@ def build(self) -> Tuple[bytes, str]:
+        definitions = " ".join(self._build_option.get_definations())
+        contents = textwrap.dedent(
+            f"""
+            cmake_minimum_required(VERSION 3.18 FATAL_ERROR)
```
It does not feel like it needs any of 3.18 feature, so why put this constraint here?

desertfire replied:
Copied from the main CMakeLists.txt in pytorch, but I agree it seems unnecessary here.

desertfire replied:
Hmm, it actually causes the cuda test fail to compile. Not sure why, but I am keeping this.

malfet commented on torch/_inductor/cpp_builder.py:
```diff
@@ -1545,3 +1547,65 @@ def build(self) -> Tuple[bytes, str]:
+            set(CMAKE_CXX_STANDARD 17)
+
+            # Set the compiler
+            set(CMAKE_CXX_COMPILER {self._compiler})
```
Why this is is needed? Cmake should be able to auto-detect compiler you are using. This will also make AOTI C++ packaging non-portable, i.e. imagine it was generated on machine with gcc-9 as host compiler, but will be compiled on machine with gcc-10

desertfire replied:
Let drop this.

malfet reviewed (approved):
Looks good to me

desertfire committed:
Update on "[AOTI] Emit a CMakeLists.txt when package_cpp_only"


Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67458526).

desertfire committed:
Update on "[AOTI] Emit a CMakeLists.txt when package_cpp_only"


Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

Differential Revision: [D67458526](https://our.internmc.facebook.com/intern/diff/D67458526)

[ghstack-poisoned]

desertfire commented:
@desertfire has imported this pull request.  If you are a Meta employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D67458526).

desertfire commented:
@pytorchbot merge

huydhn commented:
@pytorchbot revert -m 'Sorry for reverting your change but the new test is failing on ROCm' -c nosignal

inductor/test_aot_inductor_package.py::TestAOTInductorPackageCpp_cpu::test_compile_after_package [GH job link](https://github.com/pytorch/pytorch/actions/runs/12421468046/job/34685265737) [HUD commit link](https://hud.pytorch.org/pytorch/pytorch/commit/429f4cd1408b11a7b0dd10634b46b3265dc31af1)

desertfire committed:
Update on "[AOTI] Emit a CMakeLists.txt when package_cpp_only"


Summary: Emit a CMakeLists.txt with compile and link options when package_cpp_only is specified. After unzipping AOTI generated .pt2 package file, user can manually build the generated model code in their local environment.

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames chauhang aakhundov

Differential Revision: [D67458526](https://our.internmc.facebook.com/intern/diff/D67458526)

[ghstack-poisoned]

desertfire commented:
Close to reland in https://github.com/pytorch/pytorch/pull/143680
